contracts/RegisterDomain.sol

```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.19;
+
+contract RegisterDomain {
+
+    struct Domain {
+        address owner;
+    }
+
+    address public contractOwner;
+
+    mapping(string => Domain) public domains;
+    uint256 public registrationDeposit;
+    uint256 public totalDomainsRegistered = 0;
+    string[] public registeredDomains;
+
+    event DomainRegistered(string domain, address owner, uint256 amount);
+    event DomainReleased(string domain, address owner, uint256 amount);
+    event OwnerChanged(address newOwner);
+    event RegistrationDepositChanged(uint256 newDepositAmount);
+
+    constructor() {
+        contractOwner = msg.sender;
+        registrationDeposit = 1 ether;
+    }
+
+    modifier isValidDomain(string memory _domain) {
+        require(!containsDot(_domain), "Only top-level domains are allowed");
+        _;
+    }
+
+    modifier isAvailable(string memory _domain) {
+        require(domains[_domain].owner == address(0), "Domain already exists");
+        _;
+    }
+
+    modifier onlyDomainOwner(string memory _domain) {
+        require(domains[_domain].owner == msg.sender, "Only domain owner can release the domain");
+        _;
+    }
+
+    modifier onlyOwner() {
+        require(msg.sender == contractOwner, "Is not the contract owner");
+        _;
+    }
+
+    function registerDomain(string memory _domain) public payable isValidDomain(_domain) isAvailable(_domain) {
+        require(msg.value == registrationDeposit, "Deposit amount incorrect");
+        domains[_domain].owner = msg.sender;
+
+        totalDomainsRegistered += 1;
+        registeredDomains.push(_domain);
+
+        emit DomainRegistered(_domain, msg.sender, msg.value);
+    }
+
+    function releaseDomain(string memory _domain) public onlyDomainOwner(_domain) {
+        payable(msg.sender).transfer(registrationDeposit);
+        domains[_domain].owner = address(0);
+        emit DomainReleased(_domain, msg.sender, registrationDeposit);
+    }
```

Comment:
Emagin scenario:

1) initial depoisit is 1 eth
2) user A registers a domain for 1 eth
3) user B registers a domain for 1 eth
4) owner increases the registration deposit to 2 eth
5) userA releases his domain, takes 2 eth and says "thank you, userB and owner"

-2 points